src/Frame.cpp

```diff
@@ -1,227 +1,493 @@
+
+   auto cutoff = 0.833;
+   for (size_t i = 0; i < size(); i++) {
+      auto rad = guess_bonds_radius(topology_[i]).value_or(0);
+      cutoff = std::max(cutoff, rad);
+   }
+   cutoff = 1.2 * cutoff;
+
+   if (numAtoms == 0 || cutoff == 0.0) {
+      return;
+   }
+
+   //Get box's min and max
+   Vector3D min, max;
+   getMinMaxBox(positions_, numAtoms, min, max);
+
+   //Box's xyz ranges
+   float xRange = max[0] - min[0];
+   float yRange = max[1] - min[1];
+   float zRange = max[2] - min[2];
+
+   //Box's number of cells in each dim, in x and y dimensions, and in total (in x, y and z dimensions)
+   const float inversedCutoff = 1.0 / cutoff;
+   int xCells = ((int)(xRange * inversedCutoff)) + 1;
+   int yCells = ((int)(yRange * inversedCutoff)) + 1;
+   int zCells = ((int)(zRange * inversedCutoff)) + 1;
+   int xyTotalCells = yCells * xCells;
+   int totalCells = xyTotalCells * zCells;
+
+   //Check if total cells is not negative
+   if (totalCells <= 0)
+   {
+      throw error(
+         "Negative number of cells"
+      );
+      return;
+   }
+
+   //Limit number of cells in the box (to avoid memory issues)
+   if (totalCells > 40000000000)
+   {
+      throw error(
+         "Exceeded maximum cells"
+      );
+      return;
+   }
+
+   //Sort atoms in cells
+   std::vector<std::vector<int>> cells;
+   std::vector<int> numInCell;
+   std::vector<int> maxInCell;
+
+   cells.resize(totalCells);
+   numInCell.resize(totalCells);
+   maxInCell.resize(totalCells);
+
+   for (int i = 0; i < numAtoms; i++) {
+      int atomXCellNum, atomYCellNum, atomZCellNum, atomCellIndex, numOfAtomsInCell;
+
+      //Compute atom's cell index
+      Vector3D atomCoord = positions_[i];
+      atomXCellNum = (int)((atomCoord[0] - min[0]) * inversedCutoff);
+      atomYCellNum = (int)((atomCoord[1] - min[1]) * inversedCutoff);
+      atomZCellNum = (int)((atomCoord[2] - min[2]) * inversedCutoff);
+
+      atomCellIndex = atomZCellNum * xyTotalCells + atomYCellNum * xCells + atomXCellNum;
+
+      //Grow cell
+      numOfAtomsInCell = numInCell[atomCellIndex];
+      if (numOfAtomsInCell == maxInCell[atomCellIndex]) {
+         cells[atomCellIndex].resize(numOfAtomsInCell + 4);
+         maxInCell[atomCellIndex] += 4;
+      }
+
+      //Store atom index in appropriate cell
+      cells[atomCellIndex][numOfAtomsInCell] = i;
+      numInCell[atomCellIndex]++;
+   }
+
+   maxInCell.clear();
+
+   //Create the neighbor list 
+   std::vector<std::vector<int>> nbList;
+   nbList.resize(totalCells);
+   createNeighborList(nbList, xCells, yCells, zCells);
+
+   //Max number of 12 bonds per atom
+   int maxNumBonds = numAtoms * 12;
+   std::vector<int> bondResults;
+   int bondsCount = generateBonds(bondResults, totalCells, cells, numInCell, nbList, maxNumBonds, cutoff);
+
+   cells.clear();
+   nbList.clear();
+   numInCell.clear();
+
+   if (bondsCount == 0 || bondsCount > maxNumBonds)
+   {
+      if (bondsCount == 0)
+      {
+         throw error(
```

Comment:
I don't see why this should be an error, it can happen if all atoms are far appart